### PR TITLE
Compatibility with cstruct v4

### LINKED ISSUE #23: Patch dissect.etl with cstruct v.4
Branch dissect.etl with cstruct patches and dependency on >=4.0.dev <=5.0.dev

### PR DESCRIPTION
**Pull request by Miauwkeru**
Closes #23

### CODE REVIEW

File: dissect/etl/headers/event.py [hunk 1]
```diff
@@ -122,7 +123,7 @@ class EventHeaderExtendedDataItem:
         "raw_data",
     ]
 
-    def __init__(self, payload):
+    def __init__(self, payload: memoryview):
```
Comment (Schamper) [inline dissect/etl/headers/event.py:126]:
```suggestion
    def __init__(self, payload: bytes):
```
I don't think it's restricted to a memoryview?

File: dissect/etl/headers/headers.py [hunk 2]
```diff
@@ -2,14 +2,20 @@
...
 
 from dissect.cstruct.types.structure import Structure
 
 from dissect.etl.exceptions import InvalidHeaderError, InvalidMarkerError
 from dissect.etl.utils import bytes_left, c_etl_headers
 
+if TYPE_CHECKING:
+    from dissect.etl.etl import ETL
+else:
+    ETL = "ETL"
```
Comment (Schamper) [inline dissect/etl/headers/headers.py:16]: What is this necessary for?

File: dissect/etl/headers/headers.py [hunk 3]
```diff
@@ -161,7 +167,7 @@ def __repr__(self):
 class InvalidHeader(Header):
     """An invalid header."""
 
-    def __init__(self, marker, data, etl):
+    def __init__(self, marker: Marker, data: memoryview, etl: ETL):
```
Comment (Schamper) [inline dissect/etl/headers/headers.py:170]:
```suggestion
    def __init__(self, marker: Marker, data: bytes, etl: ETL):
```

File: dissect/etl/headers/headers.py [hunk 4]
```diff
@@ -185,7 +191,7 @@ class EventProperty(IntEnum):
 
 
 class MessageTraceHeader(Header):
-    def __init__(self, marker: Marker, data: memoryview, etl):
+    def __init__(self, marker: Marker, data: memoryview, etl: ETL):
```
Comment (Schamper) [inline dissect/etl/headers/headers.py:194]:
```suggestion
    def __init__(self, marker: Marker, data: bytes, etl: ETL):
```

File: dissect/etl/utils.py [hunk 5]
```diff
@@ -51,8 +51,7 @@
 #define EVENT_TRACE_TYPE_WINEVT_SEND        0x09  // Send Event (WinEvent compatible)
 #define EVENT_TRACE_TYPE_WINEVT_RECEIVE     0xF0  // Receive Event (WinEvent compatible)
 """
-c_etl_global = cstruct()
-c_etl_global.load(c_global_def)
+c_etl_global = cstruct().load(c_global_def)
```
Comment (Schamper) [inline dissect/etl/utils.py:54]:
```suggestion
c_etl_global = cstruct().load(global_def)
```

File: dissect/etl/utils.py [hunk 6]
```diff
@@ -307,8 +306,7 @@
 
 
 """
-c_etl_headers = cstruct()
-c_etl_headers.load(c_etl_definitions)
+c_etl_headers = cstruct().load(c_etl_definitions)
```
Comment (Schamper) [inline dissect/etl/utils.py:309]:
```suggestion
c_etl_headers = cstruct().load(etl_def)
```

**Review by Schamper — DISMISSED** [review]

File: dissect/etl/headers/headers.py [hunk 7]
```diff
@@ -204,7 +210,7 @@ def minimal_size(self) -> int:
     def _header_type(self) -> Structure:
```
Comment (pyrco) [inline dissect/etl/headers/headers.py:210]:
```suggestion
    def _header_type(self) -> c_etl_headers.MessageHeader:
```

Comment (pyrco) [reply dissect/etl/headers/headers.py:210]: The other `Structure` types below can also be made more specific.

Given `ErrorHeader._header_type()` returns `None`, the base class `Header._header_type()` return type should probably be changeds to `Structure | None`.

**Review by pyrco — CHANGES_REQUESTED** [review]
The `Structure` return types in `system.py` could also be made more specific.

**Review by pyrco — APPROVED** [review]

**Pull request merged**